### PR TITLE
Initial binary search

### PR DESCRIPTION
**Pull request by roderickhsiao**
@hankhsiao @imowenlee please help to review :)

### CODE REVIEW

**Comment by roderickhsiao** [comment]
@yctai

File: javascript/algorithms/binary_search.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            debug('Found match');
+            return index;
+        }
+
+        if (target[index] < element) {
+            start = index + 1;
+            debug('Swifting start index', start);
+        } else if (target[index] > element) {
+            end = index - 1;
+            debug('Swifting end index', end);
```
Comment (yctai) [inline javascript/algorithms/binary_search.js]: Looks like infinity loop will happen when element is not found.

Comment (yctai) [inline javascript/algorithms/binary_search.js]: correct myself.  The index of start keeps increasing and the index of end keeps decreasing, so start > end will happen sometime later.

The only condition for infinity loop is that if target[index] == element but target[index] !== element.   It happens when sometimes element is string and searching array is number.

Comment (roderickhsiao) [inline javascript/algorithms/binary_search.js]: good catch

**Pull request merged**